Mirror reflection

matkrak opened the pull request:
Sorry for the mess with merging branches in my own repository. I was sure that I created this PR earlier directly to your repository.

Additional info in the first commit Update profile.py

matkrak committed:
Update profile.py

1. Comments about x_at_y added (function not working as it is suppose to be)
2. Method mirror of LateralProfile updated so that now it can be given Y value (which was always the middle of section of X till now, and now is OY by default).

ad 2. Need separate method that can "fix" profile by sorting the points in it (asc X coord. by default probably)

matkrak committed:
Merge pull request #1 from matkrak/Mirror-reflection

Update profile.py

matkrak committed:
Update profile.py

grzanka commented on profile.py:
```diff
@@ -24,40 +24,40 @@ def __array_finalize__(self, obj):
+        # :param y: reference value
+        # :param reverse: boolean value - direction of lookup
+        # :return: x value corresponding to given y or NaN if not found
+        # """
```
Why docstring is commented ?

grzanka commented on profile.py:
```diff
@@ -80,6 +80,19 @@ def x_at_y(self, y, reverse=False):
+        ## Also since the function returns x value given y value,
+        ## names of variable used in loop should be y not x (in my opinion)
+
+
```
Read carefully docstring. 
We assume that the profile is first (more-less) rising then (more-less) falling down.
This function is doing lookup "from the left", checking if correspoing x is found on the (left) rising slope.
In the case of "descending profile" no such x is found and it is correct.
Try to call your function with reverse=True.

You are anyway right - better documentation is needed

matkrak committed:
Update profile.py

fix_profile method added. For now it is a method of LateralProfile, but finally would be better to make it Profile method. 
Uncommended docstrings (wasn't able to run main, instead doctests were starting, but solved it already).

matkrak commented:
If I can leave fix_profile as LateralProfile methon for now and change it do Profile method once I get more familliar with all this metadata, the last update closes #6

grzanka commented on profile.py:
```diff
@@ -23,6 +23,8 @@ def __array_finalize__(self, obj):
         self.axis = getattr(obj, 'axis', None)
 
+
+
```
Why this whitespaces ?

grzanka commented on profile.py:
```diff
@@ -150,8 +172,10 @@ def right_penumbra(self, upper=0.9, lower=0.1):
+    def mirror(self, m=0):
+        # self.y = 2*m - self.y
+        self.x = 2*m - self.x
+        self.fix_profile()
```
Do you really need to sort profile using sorting method ?
Maybe this might be helpful (a is numpy array):
a = a[::-1]

matkrak replied:
well, if given Y is not inside the domain of profile inverting the array could be used (and probably would be much more efficient then sort). On the other hand when given Y is somewhere in the domain of profile (X) it would require some extra operations, although I think it is possible and if I figure it out I will change it to use only inverting function (a[::-1])

matkrak replied:
Of course you are right, I don't know why I got confused about this one. Sure, since "breaking" the profile is done by invering the array, fixing can be done by flipping it again. Right now I am on the way to Krakow, but when I arrive I'm going to correct the code. Also I think for now I will leave fix_profile function. It is not used anywhere right now, but might be useful in the future.

grzanka commented on profile.py:
```diff
@@ -141,6 +156,13 @@ def __array_finalize__(self, obj):
 
+    def fix_profile(self, coordinate = Axis.x):
+        fun = lambda data: data[coordinate-1]
+        tempdata = [(self.x[ind], self.y[ind]) for ind in range(0,len(self.x))]
```
Two questions:
- do you know "zip" method ?
- do you know what is behind self.x and self.y (see implementation of Curve) ?

matkrak commented:
Some things came out and I have to deal with them now, I probably won't be able to push any changes today. I am going to get down to it tomorrow morning, but for now:
1) I see that zip method can be used instead of operations I did in fix_profile. That would also work for 3-D points, and as a result fix_profile function could be a method of Profile instead of LaterProfile.
2) self.x(y) contains DataSeries instance, and with respect to futher changes like introducing metadata it might occur necessary for it to have some extra fields. That makes fix_profile funcion as it is now not really universal. I probably should use zip() method.
3) What I am going to do after all is:
- rewrite fix_profile as Profile method using zip() method
- change mirror function so that it used data[::-1] method instead of data.sort()

If this is done correctly, I hope to close this issue after pushing changes.

Also thank you for a good cooperation

grzanka commented:
Ad1) and Ad2) Do you understand why I came with following implementation ?

```
    @property
    def x(self):
        return self[:, 0].view(DataSerie)

    @property
    def y(self):
        return self[:, 1].view(DataSerie)
```

Why do you want to combine x and y into 2-D array, why the object itself (self) is a 2-D (or N-D) array ?

Sorting of numpy arrays is however tricky, see this:
- http://stackoverflow.com/questions/7900171/2d-sorting-with-numpy-sort-1-row-and-have-the-other-follow-the-sorting
- http://stackoverflow.com/questions/2706605/sorting-a-2d-numpy-array-by-multiple-axes

This will lead to a new issue - considering Curve not a direct subclass of np.array, but as a numpy.recarray. Then accessing X and Y would be easier and sorting would also be easier.

grzanka commented:
You should also wait with closing issue and merging PR for my approval.

matkrak committed:
Update profile.py

mirror() method for LateralProfile improved so that it wont use .sort() method but a[::-1] operation instead.
fix_profile() which is now Profile method problems:
1) Need to find a way to copy all axis and eventual metadata (2d and 3d points, for now its only 2d)
2) Depending on where is metadata stored 2nd solution could create a mess with it.

Should I map() this values of tmp to self array somehow or using zip() method is possible here?
Or should I consider this Mirror Reflection issue closed, get down to adding metadata and then figure out how to write fix_profile method properly?

matkrak committed:
Update profile.py

Removed fix_profile() function.
Anyway I think a method like this should be provided by beprof.
Another issue and PR will be opened.

matkrak committed:
Update profile.py

Added docstring to mirror() function of LateralProfile.
You never mentioned it and I assumed atb of this issue, that mirror() should be a method of LateralProfile (not Profile).
Please let me know if I should change it.

matkrak commented:
Final version of mirror() method assuming, that it should be LateralProfile method.
Anything else should be done here?

grzanka commented on profile.py:
```diff
@@ -183,12 +203,19 @@ def plateau_to_max(self):
 
 if __name__ == '__main__':
     main()
+    
```
Why this empty line ?

grzanka commented on profile.py:
```diff
@@ -80,6 +80,18 @@ def x_at_y(self, y, reverse=False):
+        ## Check main() function with the profile i entered there.
+        ## Also since the function returns x value given y value,
+        ## names of variable used in loop should be y not x (in my opinion)
+
```
Is this comment needed ?

grzanka commented on profile.py:
```diff
@@ -150,7 +162,15 @@ def right_penumbra(self, upper=0.9, lower=0.1):
         return 0.5 * (self.x_at_y(level) + self.x_at_y(level, reverse=True))
 
-    def mirror(self):
+    def mirror(self, m=0):
```
Add doctest, in similar manner as in x_at_y. 
Each test should be one line of the code, sth like:

```
print(Profile([....]).mirror())
[...]
```

matkrak committed:
Update profile.py

1) Removed empty line at the end
2) Removed unnecessary comment
3) Docstring test, but:
    mirror() does not return anything! To make it print anything I needed 3 lines of code
    1st to create object
    2nd to apply mirror method that works "in place"
    3rd to print the result, also:
        __str__ as you defined it results in this kind of output.
        I could change this test so that it was i.e. an array of points on the output. Still need more then 1 line of code.
What should I do ?

matkrak commented:
Please take a look at docstring test I wrote and give me some feedback

grzanka commented on profile.py:
```diff
@@ -80,6 +80,7 @@ def x_at_y(self, y, reverse=False):
         # B) if y < min(self.y) then condition self.y >= y will be satisfied on first item
         # np.argmax(cond) will be equal 0,
         # to exclude situation that y_handle[0] = y we also check if y < y_handle[0]
+
```
Why this empty line ?

grzanka replied:
Remove this empty line.

grzanka commented on profile.py:
```diff
@@ -150,7 +151,25 @@ def right_penumbra(self, upper=0.9, lower=0.1):
+        Use mirror() method to get f(-x):
+        >>> foo = LateralProfile([[0, 0], [1, 1], [2, 4], [4, -1]])
+        >>> foo.mirror()
+        >>> print(foo)
```
Just a comment: **str** method will be rewritten in the future. 
From this test you can't see clearly that the profile was mirrored (print(foo) doesn't print data in the object)
Wouldn't look better if you print foo.y ?

matkrak replied:
Yes, you are right, I was just about to do this as I have in Rebinning methods docstrings tests.

matkrak committed:
Rewrote docstring test

Also decided mirror function will now return self. Now operations like
    lp1 = LateralProfile(some_array)  
    lpmirrored = lp1.mirror(some_value)
are possible and docstring looks way better

matkrak commented:
Please have a look at the last version and tell me if there is anything more I should do here?

grzanka commented:
I will download your code and let you know.

grzanka commented on profile.py:
```diff
@@ -150,8 +151,22 @@ def right_penumbra(self, upper=0.9, lower=0.1):
+        self.x = 2*m - self.x
+        self.x = self.x[::-1]
         self.y = self.y[::-1]
+        return self
```
This method is doing mirroring in-place, modifying original object. It doesn't make sense to return reference to `self` in that case. Check for example `sort` method of class `list`.

```
l = [0,2,1]
x = l.sort()
print(l)
print(x)
```

matkrak committed:
Removed return statement from LP.mirror()

You are right - return statement doesn't make sense here. But when I am trying to run docstring test as:
        >>> print((LateralProfile([[-1, 1], [0, -1], [1, 0]]).mirror()).y)
it fails with an error: AttributeError: 'NoneType' object has no attribute 'y'

Thats why I made this 3-line docstring test again. Will think about it again tomorrow.
